### PR TITLE
Editor: HTML Editor per block

### PR DESCRIPTION
**Pull request by youknowriad**
## Description
This PR explores the idea in #2794 This is largely unfinished yet but this shows how this could work. The button to switch to HTML mode is located on the menu on the right of block instead of the inspector because it's not possible to keep showing the inspector controls while in HTML mode (this is probably temporary, we need some design input here)

## Checklist:
- [x] Button to toggle the mode per block
- [x] BlockHTML form
- [x] Run the block validation when switching back to the visual mode
- [x] Make this pretty
- [x] Add unit tests

## Screenshots (jpeg or gifs if applicable):
screen shot 2017-09-26 at 14 47 27

### CODE REVIEW

**Comment by youknowriad** [comment]
@mtias Actually, I'm already using it when you're still on the visual mode, but the icon changes when you trigger the HTML mode. Probably a bad choice but you know, early iterations :)

**Comment by mcsf** [comment]
It would be cool if, when switching back to Visual, the same kind of prompts would be seen as when returning from global HTML to global Visual mode:

screen shot 2017-09-26 at 15 19 19

(For the record, currently, the changes performed in per-block HTML mode are discarded and the visual block is rendered as it previously was.)

**Comment by youknowriad** [comment]
@mcsf I was thinking it was already the case, but I think you're right, I need to run the validation manually

**Comment by mtias** [comment]
There's a case to be made that opting into the HTML mode converts it to an HTML block entirely.

**Comment by mtias** [comment]
Also, somewhat related, the invalid dialog should include an option to edit as an HTML block, not just Classic Text.

**Comment by jasmussen** [comment]
Love this feature!

Designwise I think we'll want to think about this for a bit. I'm concerned with adding still more buttons to the side. (Oh and tiny nitpick on current implementation, the hover text says "Delete the block").

For a while I've been suggesting that we implement an _overflow menu_ — that is, an ellipsis that sits at the end of every toolbar. On click it shows a dropdown with items that don't fit in the menu. This is especially important for mobile, where the toolbar might not fit. This space would be perfect for an "Edit HTML" button, I think. What do you think, @karmatosed

**Comment by youknowriad** [comment]
@jasmussen Agreed about the design issues, If we can have a mockup or I'd be to try to implement this (maybe separately)

**Comment by aduth** [comment]
Random thought: With these changes, could we potentially consider dropping the Visual / Text toggle altogether? Maybe a bit too controversial 😉

**Comment by youknowriad** [comment]
> Random thought: With these changes, could we potentially consider dropping the Visual / Text toggle altogether? Maybe a bit too controversial

I'm also in favor of this :)

**Comment by karmatosed** [comment]
> Random thought: With these changes, could we potentially consider dropping the Visual / Text toggle altogether? Maybe a bit too controversial

I would suggest usability testing on that. I am not personally against it though, it always has been confusing.

**Comment by jasmussen** [comment]
Let's step back from the cliff for a second and regroup.

There's a use case for people who never want to use the visual editor, they just want the text editor. We should let them — they might never use Gutenblocks at all, and would have no HTML comments. We should even cookie their choice, so if they are in HTML mode and reload the page, they stay there. 

For the occasions where you want to fix tag soup, it makes total sense to be able to edit on a per block basis. Having a "edit HTML" button in context of the block itself is also a delightful callback to the Visual/HTML tabs of the current editor.

As such it seems like there are some opportunities here:

a) rethinking where the editor mode switch lives. Perhaps in the ellipsis menu as suggested in another thread

b) if we can add an unread notification dot to the ellipsis menu, perhaps even a pointer bubble when new plugins add themselves there (see also #2460 ), then when a page builder installs itself, it can draw attention to the "mode switch" there

If we think this is cool, the big challenge then becomes: where do we put this html mode switch, without overloading the toolbar? Is this switch discoverable enough if we were to add an ellipsis dropdown at the end of every toolbar?

**Comment by youknowriad** [comment]
One crazy idea I was thinking of for the Text mode is using Gutenberg (or precisely a block list) also for the text mode. Let me detail this:

 - Let's say we split Gutenberg into reusable components, one of these components is the `BlockList` component where we provide an array of blocks to display and also a list of available blocks to add.

 - We use a `BlockList` with only a unique block type: `HTML block`, which means the Text mode is a list of HTML blocks, the comments stay hidden, the user can't mess with them.

 - The "serializer" for this Text mode is not the regular Gutenberg serializer, but just a simple one which concats the previous comments with the new HTML content.

**Comment by jasmussen** [comment]
I utterly love that idea, Riad.

**Comment by iseulde** [comment]
@youknowriad @jasmussen I proposed this a while ago: https://wordpress.slack.com/archives/C02QB2JS7/p1496215809472978

One advantage there too is that we can scroll to the same block the switching mode.

**Comment by jasmussen** [comment]
Here are mockups of what it would look like if the HTML mode was in the quick toolbar ellipsis menu:

edit html

edit html ellipsis toggled

I keep coming back to the quick toolbar having an ellipsis dropdown being an important task we should start work on soon. We can also choose to show as many items in the quick toolbar as there's space for, and then only "pop off" buttons to the ellipsis menu when we run out of space. In this case there would be space for the two new buttons, so it would look like this:

edit html wide quick toolbar

Here's where the global document switch would live:

edit html global switch

Note, the "footnote" and "spell checking" items are simply examples of what other elements could live in those overflow menus. 

CC: @mtias @karmatosed @youknowriad

**Comment by aduth** [comment]
>There's a use case for people who never want to use the visual editor, they just want the text editor. We should let them — they might never use Gutenblocks at all, and would have no HTML comments. We should even cookie their choice, so if they are in HTML mode and reload the page, they stay there.

Couldn't these users just insert a single HTML block into the editor and have-at-it?

File: editor/block-settings-menu/index.js
Comment (aduth) [inline editor/block-settings-menu/index.js]: TIL "textual" is a word - still think "HTML" or "text" might be clearer.

File: editor/reducer.js
Comment (aduth) [inline editor/reducer.js]: Tab one more.

File: editor/modes/visual-editor/block-html.js
Comment (aduth) [inline editor/modes/visual-editor/block-html.js]: Would we need to update `attributes` state here as well?

File: editor/modes/visual-editor/block-html.js
Comment (aduth) [inline editor/modes/visual-editor/block-html.js]: Why do we need to maintain `attributes` in state?

File: editor/reducer.js
Comment (aduth) [inline editor/reducer.js]: Maybe a destructure could help clear this up a bit:

```js
export function blocksMode( state = {}, action ) {
	const { type, uid } = action;
	if ( type === 'TOGGLE_BLOCK_MODE' ) {
		return {
			...state,
			[ uid ]: state[ uid ] && state[ uid ] === 'html' ? 'visual' : 'html',
		};
	}

	return state;
}
```

Or inside the condition if we want to avoid assignment for unhandled actions:


```js
export function blocksMode( state = {}, action ) {
	if ( action.type === 'TOGGLE_BLOCK_MODE' ) {
		const { uid } = action;
		return {
			...state,
			[ uid ]: state[ uid ] && state[ uid ] === 'html' ? 'visual' : 'html',
		};
	}

	return state;
}
```

**Review by aduth — DISMISSED** [review]
Strange behavior I'm seeing:

1. Navigate to Gutenberg > Demo
2. Switch the first Cover Image block to HTML mode
3. Focus in the block
4. Focus out of the block

The block changes its HTML inexplicably, resetting its content to the default state.

File: editor/modes/visual-editor/block-html.js
Comment (youknowriad) [reply editor/modes/visual-editor/block-html.js]: The idea is:

 - don't save the updated HTML (which means parse the attributes) on every key stroke (weird things happen if you do so :) without mentioning the performance reasons) and only persist these changes onBlur
 - If the attributes change externally, this means something else changed the current block,  will have to recompute the `HTML` again.
 - Each time the text mode changes the attributes we store them in state to compare with the new prop.

If you have another suggestion to achieve this, I'm all for it.

File: blocks/api/serializer.js
Comment (aduth) [inline blocks/api/serializer.js]: Can we add a DocBlock, particularly because it's not obvious how this function is distinct from the other four "get me some content" functions in this file.

File: editor/modes/visual-editor/block-html.js
Comment (aduth) [reply editor/modes/visual-editor/block-html.js]: >don't save the updated HTML (which means parse the attributes) on every key stroke (weird things happen if you do so :) without mentioning the performance reasons) and only persist these changes onBlur

This is fine, I understand we don't want to update the block until blur, so storing initial HTML makes sense. Doesn't seem to imply we need `attributes` to do so, though.

>If the attributes change externally, this means something else changed the current block, will have to recompute the HTML again.

>Each time the text mode changes the attributes we store them in state to compare with the new prop.

We can compare `isEqual( this.props.attributes, nextProps.block.attributes )` in `componentWillReceiveProps`

File: editor/modes/visual-editor/style.scss
Comment (aduth) [inline editor/modes/visual-editor/style.scss]: Minor: In aiming to avoid the tag name selector, and since we don't apply any styles to `blocks-visual-editor__block-html`, maybe we skip applying this class name to the root element and instead target the textarea with a classname `.blocks-visual-editor__block-html-textarea`?

**Comment by iseulde** [comment]
@jasmussen Honestly my first thought with the ellipsis (both at block level and document level) was... not another hidden area. We already have the inspector, which kind of acts as a place were all the advanced stuff goes. Also for the one at the top next to publish, is that not staff that could go in the document inspector? The generic cog and ellipsis next to each other feel quite strange. What doe those even mean? What's the difference? I understand the need for something at the inline block level, where we need a solution for overflow, I'm still wondering though if we can't just put an ellipsis there that would open up the inspector (and get rid of the other one).

File: editor/modes/visual-editor/block-html.js
Comment (youknowriad) [reply editor/modes/visual-editor/block-html.js]: you're right :+1:

File: editor/block-settings-menu/index.js
Comment (aduth) [reply editor/block-settings-menu/index.js]: >TIL "textual" is a word - still think "HTML" or "text" might be clearer.

Still think this ought to be addressed

File: editor/modes/visual-editor/block-html.js
Comment (aduth) [inline editor/modes/visual-editor/block-html.js]: Why do we need `omit` here if we proceed to assign in `sourcedAttributes`? In other words, the keys of `sourcedAttributes` will already override those from `this.props.block.attributes`, so the omit is unnecessary.

**Review by aduth — COMMENTED** [review]
Testing this again, I'm still having some trouble:

- In demo, switching the first Cover Image block to HTML, I proceed to remove the "has-background-dim" class from the HTML. When I switch back to visual mode, it prompts me that the block has been modified externally. I think I understand this to be caused by an incompatibility between block comments and the markup of the block, but from a user's perspective this is not very obvious. I'm not really sure how we'd want to reconcile this (include the comments in the HTML?)
- After the invalid block prompt is shown, I wanted to switch back to HTML mode and fix the issue, but clicking "HTML" does nothing at this point. Should we allow a user to switch back to HTML mode to fix the block themselves?

File: editor/block-settings-menu/index.js
Comment (youknowriad) [reply editor/block-settings-menu/index.js]: sorry forgot about this, updating

Comment (youknowriad) [reply editor/block-settings-menu/index.js]: going for "text" for consistency with the global mode switcher

File: editor/modes/visual-editor/block-html.js
Comment (youknowriad) [reply editor/modes/visual-editor/block-html.js]: I was mistakenly thinking `Object.assign({ a: 'value' }, { a: undefined });` would return `{ a: 'value' }` but I was wrong.

**Comment by youknowriad** [comment]
@aduth Good questions

> After the invalid block prompt is shown, I wanted to switch back to HTML mode and fix the issue, but clicking "HTML" does nothing at this point. Should we allow a user to switch back to HTML mode to fix the block themselves?

Sounds like a good idea to me, this and showing a nudge indicating if the block is valid or not somewhere when in HTML mode. Thoughts @jasmussen?

> In demo, switching the first Cover Image block to HTML, I proceed to remove the "has-background-dim" class from the HTML. When I switch back to visual mode, it prompts me that the block has been modified externally. I think I understand this to be caused by an incompatibility between block comments and the markup of the block, but from a user's perspective this is not very obvious. I'm not really sure how we'd want to reconcile this (include the comments in the HTML?)

I think the "isValid" nudge might help a bit here.

**Comment by mtias** [comment]
If we nudge when something is invalid, we ought to also give the option to convert to a plain HTML block. (I somehow expect this to be the most common case here.)

**Comment by karmatosed** [comment]
I have to admit visually this is a little 'jarring' and feels easy to miss/confuse.

2017-10-03 at 21 56

It's worth adding, I think we need to explore ... as an option soon for the toolbar anyway, I am not sure this is something to hide there, no. But it's likely to appear and very likely on mobile.

I like the idea of a 'tools' icon maybe for checking spelling and showing html. However we do that. One idea could be have a tools icon. I think top bar feels right for this for the entire document though.

This is just a fast exploration but what about:

2017-10-03 at 21 56

**Comment by jasmussen** [comment]
I have a very hard time with us putting a single other button on the side of this box. I've already been arguing we should find a way to consolidate this a bit (and I'm aware of how difficult this is), but having _three_ buttons on the side I don't think can work simply for the fact that it wouldn't "fit" at the side of a single line paragraph block. 

At this point I think this mockup reads the best:



Only two problems with that one, as far as I can tell, 1 — it's perhaps a slight stretch to call "Edit HTML" a "formatting" option, but it _is_ a block option. And 2 — it doesn't scale well to mobile unless we implement the overflow ellipsis menu so items can pop off the quick toolbar like they do in Google Docs. I know this is controversial, but for all I care it doesn't have to be an ellipsis — it could be a text label, "More", exactly like Docs:

docs

**Comment by jasmussen** [comment]
Here are some more mockups for how to handle overflow in the quick toolbar.

Desktop breakpoint:

desktop toolbar overflow

Mobile breakpoint version 1, toolbar scrolls:

mobile overflow scroll
mobile overflow scrolled

Mobile breakpoint version 2, "More" button appears (like Docs):

mobile overflow more button

mobile overflow more popout

Pros and cons to all of this:

- The scrollable quick toolbar won't work on the desktop. It also has challenges with regards to quick toolbar dropdowns, like the switcher.
- The More button I personally like, but it's probably tricky coding.

Whatever we decide on this, I feel like we need a version of the above. Keeping the quick toolbar slim is a noble goal which we absolutely shouldn't drop. But right now we are putting things on the side of the block, or in the inspector (which users can toggle off) — this is an imbalance, I feel.

**Comment by karmatosed** [comment]
I second that the size we have at the side just doesn't really work, I was fitting the best I could and my visuals show it's not the best we can do. I would second having a 'more' option both globally and per block. My only caution is mixing '...' and 'more', I feel it would need to have the same language perhaps?

**Comment by youknowriad** [comment]
One drawback of using the block toolbar for this, is that the block toolbar buttons disappears when we switch to the HTML mode, so we'll have a unique button on the toolbar to move back to the visual mode.

Also, Moving these icons to the toolbar and handling the overflow in mobile seems like a big task on its own. Should I try to solve this here or leave it to its own issue?

**Comment by jasmussen** [comment]
I really love this feature, and how it improves the handling of this all. But I'm a bit worried about us adding a ton of UI with the hopes that we'll fix it later on. It's not the hill I'll die on, if there's a strong desire to move forward, but I think it'd be good to keep thinking a bit about the UI, perhaps a simple solution can present itself. Perhaps there's preceeding art we can look at. I mean, the current editor has tabs for Visual and Text. Though not the greatest UI (and probably can't form the basis for what we're doing here) it's an example. I'll look around for a bit and see if another solution pops up.

**Comment by mtias** [comment]
@jasmussen what do you think of combining the trash icon, the cog settings, and this new html mode under an ellipsis—but keeping the ellipsis in the right floating side instead of the toolbar?

**Comment by jasmussen** [comment]
> what do you think of combining the trash icon, the cog settings, and this new html mode under an ellipsis—but keeping the ellipsis in the right floating side instead of the toolbar?

Love it.

But I never thought that would fly.

**Comment by youknowriad** [comment]
After #2884 merge, I updated this and it's in the block's menu right now.

Also, What about the "invalid" nudge, any design thoughts or mockups on this @jasmussen?

**Comment by jasmussen** [comment]
> Also, What about the "invalid" nudge, any design thoughts or mockups on this @jasmussen?

I think that is probably best explored separately. But it could be something in this vein:

screen shot 2017-10-09 at 15 18 08

CC: @mtias as I know he had a bunch of thoughts on that.

**Comment by jasmussen** [comment]
> Nice design, so do you feel this is ready to go? what's left here?

I feel like this is ready to go, yes. A nice enhancement could be to have the "html mode" button have a "toggled" state — just like if it were a "bold" button. That is, dark square background, and white text on it. I was also thinking instead of having Trash and Configure float on the side, we could put them in a popout menu. But this is an enhancement we can do later on, I feel.

Thoughts, @mtias ? I think it's 👍 👍

**Comment by karmatosed** [comment]
👍 from me and :shipit: :)

**Comment by afercia** [comment]
From an accessibility perspective, I have no objections if you want to implement an HTML mode that is basically a single contenteditable block.

I'd just like to remind you all that a "text mode" with a simple textarea is a must-have for accessibility, as it's the only guarantee it can be operated by all users with any device or assistive technology.

**Comment by youknowriad** [comment]
Merging this, should we open an issue for the updated design

**Comment by jasmussen** [comment]
> Merging this, should we open an issue for the updated design

I'm working on some enhancements here, will open a ticket as soon as I can.

**Pull request merged**
